hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,243 @@
+      final PlacementPolicy containerPlacement, final ConfigurationSource conf,
+      NodeManager nodeManager) {
+    this.containerPlacement = containerPlacement;
+    this.currentContainerSize = (long) conf
+        .getStorageSize(ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE,
+            ScmConfigKeys.OZONE_SCM_CONTAINER_SIZE_DEFAULT, StorageUnit.BYTES);
+    this.nodeManager = nodeManager;
+  }
+
+  /**
+   * Identify a new set of datanode(s) to reconstruct the container and form the
+   * SCM command to send it to DN. In the case of decommission, it will just
+   * generate the replicate commands instead of reconstruction commands.
+   *
+   * @param replicas - An instance of ContainerReplicaCount, containing the
+   *                   current replica count and inflight adds and deletes
+   * @param pendingOps - Inflight replications and deletion ops.
+   * @param result - Health check result.
+   * @param remainingMaintenanceRedundancy - represents that how many nodes go
+   *                                      into maintenance.
+   * @return Returns the key value pair of destination dn where the command gets
+   * executed and the command itself.
+   */
+  @Override
+  public Map<DatanodeDetails, SCMCommand> processAndCreateCommands(
+      final Set<ContainerReplica> replicas,
+      final List<ContainerReplicaOp> pendingOps,
+      final ContainerHealthResult result,
+      final int remainingMaintenanceRedundancy) {
+    ContainerInfo container = result.getContainerInfo();
+    ECReplicationConfig repConfig =
+        (ECReplicationConfig) container.getReplicationConfig();
+    final ECContainerReplicaCount replicaCount =
+        new ECContainerReplicaCount(container, replicas, pendingOps,
+            remainingMaintenanceRedundancy);
+
+    ContainerHealthResult currentUnderRepRes = ecContainerHealthCheck
+        .checkHealth(container, replicas, pendingOps,
+            remainingMaintenanceRedundancy);
+
+    LOG.debug("Handling under-replicated EC container: {}", container);
+    if (currentUnderRepRes
+        .getHealthState() != ContainerHealthResult.HealthState
+        .UNDER_REPLICATED) {
+      LOG.info("The container {} state changed and it's not in under"
+              + " replication any more. Current state is: {}",
+          container.getContainerID(), currentUnderRepRes);
+      return null;
+    }
+    ContainerHealthResult.UnderReplicatedHealthResult containerHealthResult =
+        ((ContainerHealthResult.UnderReplicatedHealthResult)
+            currentUnderRepRes);
+    if (containerHealthResult.isSufficientlyReplicatedAfterPending()) {
+      LOG.info("The container {} with replicas {} is sufficiently replicated",
+          container.getContainerID(), replicaCount.getReplicas());
+      return null;
+    }
+    if (replicaCount.unRecoverable()) {
+      LOG.warn("The container {} is unrecoverable. The available replicas" +
+          " are: {}.", container.containerID(), replicaCount.getReplicas());
+    }
```

Comment:
Is it intentional that the method proceeds for unrecoverable case?

Reply:
Good catch. I did return empty map now.